library/pkcs11_client.c

```diff
@@ -0,0 +1,527 @@
+    mbedtls_pk_type_t can_do;
+    CK_ATTRIBUTE attributes[] = {
+        {CKA_CLASS, &public_key_class, sizeof( public_key_class )},
+        {CKA_KEY_TYPE, &public_key_type, sizeof( public_key_type )},
+    };
+    CK_RV rv;
+    uint16_t key_size = 0;
+
+    rv = C_GetAttributeValue( hSession, hPublicKey,
+                              attributes, ARRAY_LENGTH( attributes ) );
+    if( rv != CKR_OK )
+        return( pkcs11_err_to_mbedtls_pk_err( rv ) );
+    if( public_key_class != CKO_PUBLIC_KEY )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    if( hPrivateKey != CK_INVALID_HANDLE )
+    {
+        attributes[0].pValue = &private_key_class;
+        attributes[1].pValue = &private_key_type;
+        rv = C_GetAttributeValue( hSession, hPrivateKey,
+                                  attributes, ARRAY_LENGTH( attributes ) );
+        if( rv != CKR_OK )
+            return( pkcs11_err_to_mbedtls_pk_err( rv ) );
+        if( private_key_class != CKO_PRIVATE_KEY )
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+        if( public_key_type != private_key_type )
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+    }
+
+    switch( public_key_type ) {
+#if defined(MBEDTLS_ECDSA_C)
+    case CKK_ECDSA:
+        can_do = MBEDTLS_PK_ECKEY;
+        {
+            unsigned char ecParams[16];
+            mbedtls_asn1_buf params_asn1;
+            mbedtls_ecp_group_id grp_id;
+            const mbedtls_ecp_curve_info *curve_info;
+            attributes[0].type = CKA_EC_PARAMS;
+            attributes[0].pValue = ecParams;
+            attributes[0].ulValueLen = sizeof( ecParams );
+            rv = C_GetAttributeValue( hSession, hPrivateKey, attributes, 1 );
+            if( rv != CKR_OK )
+                return( pkcs11_err_to_mbedtls_pk_err( rv ) );
+            params_asn1.tag = ecParams[0];
+            params_asn1.len = ecParams[1];
+            params_asn1.p = ecParams + 2;
+            if( mbedtls_oid_get_ec_grp( &params_asn1, &grp_id ) != 0 )
+                return( MBEDTLS_ERR_PK_UNKNOWN_NAMED_CURVE );
+            curve_info = mbedtls_ecp_curve_info_from_grp_id( grp_id );
+            if( curve_info == NULL )
+                return( MBEDTLS_ERR_PK_UNKNOWN_NAMED_CURVE );
+            key_size = curve_info->bit_size;
+        }
+        break;
+#endif /* MBEDTLS_ECDSA_C */
+    default:
+        can_do = MBEDTLS_PK_OPAQUE;
+        break;
+    }
+
+    {
+        int ret = mbedtls_pk_setup( ctx, &mbedtls_pk_pkcs11_info );
+        if( ret != 0 )
+            return( MBEDTLS_ERR_PK_ALLOC_FAILED );
+    }
+    {
+        mbedtls_pk_pkcs11_context_t *pkcs11_ctx = ctx->pk_ctx;
+        pkcs11_ctx->key_type = can_do;
+        pkcs11_ctx->bit_length = key_size;
+        pkcs11_ctx->hSession = hSession;
+        pkcs11_ctx->hPublicKey = hPublicKey;
+        pkcs11_ctx->hPrivateKey = hPrivateKey;
+    }
+    return( 0 );
+}
+
+#if defined(MBEDTLS_RSA_C) || defined(MBEDTLS_ECDSA_C)
+static int mpi_to_ck( const mbedtls_mpi *mpi,
```

Comment:
Our functions return 0 on success, and a negative value on failure, see "return type" in https://tls.mbed.org/kb/development/mbedtls-coding-standards Please change this one to follow our usual convention.

(I know this sounds like nitpicking, but there has been actual bugs in security software (GnuTLS IIRC) due to a  confusion about whether 0 means success of failure. Let's avoid that by being consistent everywhere.)

Reply:
You're right. This function should propagate the error from `mbedtls_mpi_write_binary` or return 0 for success.

Reply:
Fixed.